### DISCUSSION TITLE
atomic_ref for vector types

### DISCUSSION THREAD

**Discussion by jinz2014** [intel/llvm#13345] [Q&A]
Can you please suggest some solutions when the template parameter is char, ushort, int4, long4, etc for atomic_ref  ?   Thanks.

Reference
https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html

The template parameter T must be one of the following types:

int,

unsigned int,

long,

unsigned long,

long long,

unsigned long long,

float, or

double.

**Comment by AlexeySachkov** [comment]
Hi @jinz2014,

For vectors you can create a wrapper function/class which simply performs per-element operation. `vec::operator[]` returns a reference, so it should be good enough to be passed to `atomic_ref`.

We have a proposed extension to enable `atomic_ref` for 2-byte data types: [sycl_ext_oneapi_atomic16](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_atomic16.asciidoc), but it is not implemented yet.

I'm not aware of any extensions for 1-byte types like `char`. I'm not sure if there is a good workaround except for changing data types you are using so that you can have atomic operations on them.
